### PR TITLE
Add exclude_none option

### LINKED ISSUE #587: add skip_none argument
# Question / Feature Request 

Suppose that I have a class like this:

```
class Jake(BaseModel):
    foo: Optional[str]
```

When I serialize it, I only want to serialize optional values when they are not null.  The `skip_defaults` flag does something _similar_ to what I am looking for but has this weird case:

```
>>> j = Jake()
>>> 
>>> # This is the behavior I expect: 
... j.dict(skip_defaults=True)
{}
>>> 
>>> # Set the foo property: 
>>> j.foo = "baz"
>>> 
>>> # This is also the behaviour I expect: 
... j.dict(skip_defaults=True)
{'foo': 'baz'}
>>> 
>>> 
>>> # Unfortunately, this is not the behavior that I expect:
... j.foo = None
>>> j.dict(skip_defaults=True)
{'foo': None}
```

---

Right now after I serialize to dict, I walk the dict and remove all of the `None` values recursively, but I would rather not have to do this.  Is this built into Pydantic?  Is the behavior that I am seeing a bug?

### PR DESCRIPTION
**Pull request by niknetniko**
## Change Summary

- Add `exclude_none` option to `dict()` and friends.

The original issue talked about `skip_none`, but since all options are now in the `exclude_x` format, that is also what I've done.

## Related issue number

Resolves #587

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -640,15 +649,18 @@ def _iter(
...
                     v,
                     to_dict=to_dict,
                     by_alias=by_alias,
                     include=value_include and value_include.for_element(k),
                     exclude=value_exclude and value_exclude.for_element(k),
                     exclude_unset=exclude_unset,
                     exclude_defaults=exclude_defaults,
+                    exclude_none=exclude_none,
                 )
+                if (exclude_none and value is not None) or not exclude_none:
```
Comment (samuelcolvin) [inline pydantic/main.py]:
```suggestion
                if not (exclude_none and value is None):
```
I think slightly clearer and more efficient.

File: tests/test_edge_cases.py [hunk 2]
```diff
@@ -1303,3 +1303,89 @@ class MyModel(BaseModel):
...
+    assert m.dict(exclude_none=True) == {'c': 5, 'd': 2, 'e': {'a': 0, 'b': 1}}
+    assert dict(m) == {'c': 5, 'd': 2, 'e': {'a': 0, 'b': 1}, 'f': None}
+
+    m = ModelB(c=5, e={'b': 20}, f='test')
+    assert m.dict() == {'c': 5, 'd': 2, 'e': {'a': None, 'b': 20}, 'f': 'test'}
+    assert m.dict(exclude_none=True) == {'c': 5, 'd': 2, 'e': {'b': 20}, 'f': 'test'}
+    assert dict(m) == {'c': 5, 'd': 2, 'e': {'a': None, 'b': 20}, 'f': 'test'}
+
+
+def test_dict_exclude_none_populated_by_alias():
```
Comment (samuelcolvin) [inline tests/test_edge_cases.py]: this name is confusing, but also I think the test is unnecessary.

File: tests/test_edge_cases.py [hunk 3]
```diff
@@ -1303,3 +1303,89 @@ class MyModel(BaseModel):
...
+            allow_population_by_field_name = True
+
+    m = MyModel(alias_a='a')
+
+    assert m.dict(exclude_none=True) == {'a': 'a'}
+    assert m.dict(exclude_none=True, by_alias=True) == {'alias_a': 'a'}
+    assert m.dict(by_alias=True) == {'alias_a': 'a', 'alias_b': None}
+
+
+def test_dict_exclude_none_populated_by_alias_with_extra():
```
Comment (samuelcolvin) [inline tests/test_edge_cases.py]: Best to test just one thing at a time, e.g. here just `test_exclude_none_with_extra`

File: tests/test_edge_cases.py [hunk 4]
```diff
@@ -1303,3 +1303,89 @@ class MyModel(BaseModel):
...
+            extra = 'allow'
+
+    m = MyModel(alias_a='a', c='c')
+
+    assert m.dict(exclude_none=True) == {'a': 'a', 'c': 'c'}
+    assert m.dict(exclude_none=True, by_alias=True) == {'alias_a': 'a', 'c': 'c'}
+    assert m.dict(by_alias=True) == {'alias_a': 'a', 'alias_b': None, 'c': 'c'}
+
+
+def test_dict_exclude_none_populated_with_extra():
```
Comment (samuelcolvin) [inline tests/test_edge_cases.py]: I don't think both this test and the above are required.

File: tests/test_edge_cases.py [hunk 5]
```diff
@@ -1303,3 +1303,89 @@ class MyModel(BaseModel):
 
     m = MyModel(foo={'x': 'a', 'y': None})
     assert m.foo == {'x': 'a', 'y': None}
+
+
+def test_exclude_none_dict():
```
Comment (samuelcolvin) [inline tests/test_edge_cases.py]:
```suggestion
def test_exclude_none():
```

File: tests/test_edge_cases.py [hunk 6]
```diff
@@ -1303,3 +1303,89 @@ class MyModel(BaseModel):
...
+    class MyModel(BaseModel):
+        a: Optional[int] = None
+        b: int = 2
+
+    m = MyModel(a=5)
+    assert m.dict(exclude_none=True) == {'a': 5, 'b': 2}
+
+    m = MyModel(b=3)
+    assert m.dict(exclude_none=True) == {'b': 3}
```
Comment (samuelcolvin) [inline tests/test_edge_cases.py]:
```suggestion
    assert m.dict(exclude_none=True) == {'b': 3}
    assert m.json(exclude_none=True) == '{"b": 3}'
```

File: tests/test_edge_cases.py [hunk 7]
```diff
@@ -1303,3 +1303,89 @@ class MyModel(BaseModel):
...
+
+    m = MyModel()
```
Comment (samuelcolvin) [inline tests/test_edge_cases.py]: This case doesn't test anything, can be removed.

**Comment by niknetniko** [comment]
Thanks for the speedy review; I've applied the suggestions.

**Comment by samuelcolvin** [comment]
awesome, thank you very much.

**Pull request merged**